core/src/main/scala/org/apache/spark/scheduler/SparkListener.scala

```diff
@@ -291,6 +297,16 @@ private[spark] trait SparkListenerInterface {
   def onBlockUpdated(blockUpdated: SparkListenerBlockUpdated): Unit
 
   /**
+   * Called when a speculative task is submitted
+   */
+  def onSpeculativeTaskSubmitted(speculativeTask: SparkListenerSpeculativeTaskSubmitted): Unit
+
+  /**
+   * Called when an extra executor is needed
+   */
+  def onExtraExecutorNeeded(): Unit
```

Comment:
do you know how executor allocation manager adjust the number of executors currently? can we follow it instead of hacking the `SparkListener`?